text/0000-node-reactions.md

```diff
@@ -0,0 +1,208 @@
+    2.2 If currentNode is replacing an existing child node of the parent, then the actors are the existing child node and its subtree, the currentNode and its subtree.
+    2.3 If currentNode's subtree is being mutated without changing the placement of the current node itself, then the actors are all the child nodes of currentNode and their subtree.
+3. Identify the order of traversal
+    3.1 Predetermine the order in which the hooks will be invoked.
+    3.2 Once a hook is qualified, it will be invoked regardless of any mutations that occur during invocation of prior hooks.
+4. If the same node qualifies for more than one lifecycle event e.g, appendChild(existingConnectedNode)
+    4.1 Then process all lifecycle events for the given node
+    4.2 The order of lifecycle events is: disconnectedCallback, connectedCallback
+    4.3 After all hooks for all events have been invoked, move on to other nodes as per order of traversal.
+5. If an existing node is being replaced by a new node
+    5.1 Process lifecycle events for existing node(s)
+    5.2 Next, process lifecycle events for new node(s)
+6. If the currentNode is a custom element, then the order of processing is as follows(TODO - provide link to native custom elements example)
+    6.1 The host element is processed first
+    6.2 shadowRoot and nodes in the shadow tree next
+    6.3 Nodes in the light DOM
+```
+
+### What browser APIs control a node's lifecycle?
+
+#### DOM APIs that connect a node
+
+```js
+Node.appendChild
+Node.insertBefore
+Node.replaceChild
+
+ParentNode.prepend
+ParentNode.append
+
+Element.insertAdjacentElement
+
+ChildNode.after
+ChildNode.before
+ChildNode.replaceWith
+
+Range.insertNode
+
+```
+
+#### DOM APIs that disconnect a node
+
+```js
+ChildNode.after
+ChildNode.before
+ChildNode.remove
+ChildNode.replaceWith
+
+Element.insertAdjacentElement
+Element.innerHTML
+Element.outerHTML
+
+Node.appendChild
+Node.insertBefore
+Node.nodeValue
+Node.replaceChild
+Node.removeChild
+Node.textContent
+
+Range.insertNode
+Range.deleteContents
+Range.extractContents
+Range.surroundContents
+```
+
+#### Reactions for various APIs
+Here are some sample reactions for the APIs listed above.
+- Node.appendChild: 
+    - If the childNode has an existing parent and the parent is connected, invoke disconnectedCallback for the childNode and its subtree.
+    - If new parent is connected, invoke connectedCallback for the childNode and its subtree
+
+- Node.insertBefore: Same steps as Node.appendChild
+
+- Node.replaceChild
+    - if the parent node is connected
+        - invoke disconnectedCallback for the existing child and its subtree
+        - invoke disconnectedCallback for the new child and its subtree
+
+# Drawbacks
+
+Why should we *not* do this? Please consider:
+
+- Performance: Traversing the dom is not cheap. We need to think about memoization techniques to prevent repetitive traversal. This is of significant importance since the invocations are synchronous in nature.
```

Comment:
Before merging this RFC I would be interested to measure the performance impact of such library on an existing application. Depending on the results we sill be able to better understand the trade-offs here.

Reply:
it is important to notice that today, LWC is patching the same APIs, this is just a more complete package.